### PR TITLE
Adds experimental LightningQubit2 (new device API) [basic structure]

### PR DESCRIPTION
**Pull request by AmintorDusko**
**Context:** This PR implements an experimental Lightning qubit following the new device API prescription.
This will be a base branch that will aggregate smaller PRs implementing the functionalities in PR #425.

**Description of the Change:**
Creates the new device in an experimental directory, and adds tests in a tests/experimental directory.

**Review Instruction:**
The original PR was separated into 4 parts:

- A basic device structure: PR #432.
- Execute pipeline: PR #433.
- Preprocess functionality: #434.
- Adjoint Jacobian pipeline: #435 

PR #432, will serve as a base where all PRs will be merged before merging it to the master.
I suggest reviewing #433 and #434 before #435 because there is a lot of overlap between them.
#435 is better to be reviewed only after #433 and #434 are merged to #432.

### CODE REVIEW

File: pennylane_lightning/experimental/lightning_qubit_2.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            is_single_circuit = False
+            if isinstance(circuits, QuantumScript):
+                is_single_circuit = True
+                circuits = [circuits]
+
+            if self.tracker.active:
+                self.tracker.update(batches=1, executions=len(circuits))
+                self.tracker.record()
+
+            results = tuple(simulate(c) for c in circuits)
```
Comment (mlxd) [inline pennylane_lightning/experimental/lightning_qubit_2.py:129]: The nice thing about here is that we can choose to use Ray, Dask, `concurrent.futures` compatible context managers and the like as a follow-up extension.

File: pennylane_lightning/experimental/lightning_qubit_2.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                circuit (QuantumTape): An optional circuit to check derivatives support for.
+
+            Returns:
+                Bool: Whether or not a derivative can be calculated provided the given information
+            """
+            if (
+                execution_config.gradient_method != "adjoint"
+                or execution_config.derivative_order != 1
+            ):
+                return False
```
Comment (mlxd) [inline pennylane_lightning/experimental/lightning_qubit_2.py:152]: I think all good for now. We can probably update this later to allow custom derivative operations to be registered and validated, with some extra list/dictionary storage on the device as needed.

File: tests/experimental/test_lightning_qubit_2.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+from pennylane.devices.experimental import ExecutionConfig
+from pennylane_lightning.experimental import LightningQubit2
+
+from pennylane_lightning.experimental.lightning_qubit_2 import CPP_BINARY_AVAILABLE
+
+
+@pytest.mark.skipif(not CPP_BINARY_AVAILABLE, reason="Lightning binary required")
+def test_name():
+    """Tests the name of LightningQubit2."""
+    assert LightningQubit2().name == "lightning.qubit.2"
```
Comment (mlxd) [inline tests/experimental/test_lightning_qubit_2.py:29]: Tying this into the question about static method from earlier.

File: tests/experimental/test_lightning_qubit_2.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+
+@pytest.mark.skipif(CPP_BINARY_AVAILABLE, reason="Only when there is no binary")
+def test_name_no_binary():
+    """Tests that we are offloading to the default qubit."""
+    assert LightningQubit2().name == "default.qubit.2"
```
Comment (mlxd) [inline tests/experimental/test_lightning_qubit_2.py:35]: And here too.

File: tests/experimental/test_lightning_qubit_2.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    with pytest.raises(NotImplementedError):
+        dev.compute_jvp(qml.tape.QuantumScript(), (10, 10))
+
+    with pytest.raises(NotImplementedError):
+        dev.execute_and_compute_jvp(qml.tape.QuantumScript(), (10, 10))
+
+
+@pytest.mark.skipif(not CPP_BINARY_AVAILABLE, reason="Lightning binary required")
+def test_no_vjp_functionality():
+    """Test that vjp is not supported on LightningQubit2."""
```
Comment (mlxd) [inline tests/experimental/test_lightning_qubit_2.py:62]: I assume this comes in one of the follow-up PRs?

File: pennylane_lightning/experimental/lightning_qubit_2.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        A device that interfaces with C++ to perform fast linear algebra calculations.
+
+        Use of this device requires pre-built binaries or compilation from source. Check out the
+        :doc:`/installation` guide for more details.
+
+        Args:
+            c_dtype: Datatypes for statevector representation. Must be one of ``np.complex64`` or ``np.complex128``.
+        """
+
+        def __init__(self, c_dtype=np.complex128):
```
Comment (mlxd) [inline pennylane_lightning/experimental/lightning_qubit_2.py:53]: While we don't need to solve this now, I think having the option of a stateful memory block as part of the device as a config option to be fed in from elsewhere, could be useful. I still prefer the allocate-and-run strategy, but that may not always be preferential. Let's come back to it another time.

**Review by mlxd — COMMENTED** [review]
Nice work --- a few questions, but nothing major from me here.

File: pennylane_lightning/experimental/lightning_qubit_2.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            self.C_DTYPE = c_dtype
+            if self.C_DTYPE not in [np.complex64, np.complex128]:
+                raise TypeError(f"Unsupported complex Type: {c_dtype}")
+
+            super().__init__()
+
+        @property
+        def name(self):
+            """The name of the device."""
+            return "lightning.qubit.2"
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply pennylane_lightning/experimental/lightning_qubit_2.py:63]: I decided to follow `default.qubit.2` implementation for this.
But, of course, we can make it static if you prefer, just let me know.

File: tests/experimental/test_lightning_qubit_2.py (hunk 5 above)
Comment (AmintorDusko) [reply tests/experimental/test_lightning_qubit_2.py:62]: This was on the nice-to-have list. 
I didn't have time to do that last quarter.
Let me know, if you want me to implement it.

**Pull request closed without merging**